gem/optimise.py

```diff
@@ -172,6 +172,17 @@ def _constant_fold_zero_literal(node, self):
         return node
 
 
+@_constant_fold_zero.register(ListTensor)
+def _constant_fold_zero_listtensor(node, self):
+    new_children = list(map(self, node.children))
+    if all(isinstance(nc, Zero) for nc in new_children):
+        return Zero(node.shape)
+    elif all(nc == c for nc, c in zip(new_children, node.children)):
+        return node
+    else:
+        return node.reconstruct(*new_children)
+
+
```

Comment:
I think we need this so that `ListTensor.reconstruct([Zero(), Zero()])` will not return `ListTensor(array[0., 0.])`, which would have missing indices.

Reply:
We want to postpone to constant-folding of zeros to late in the generation pass, so eager transformation of `ListTensor([Zero(), Zero()])` to `Zero((2,))` would be undesirable.